// @sonolus/express/dist/routes/items/details.js
import { Icon, Text } from '@sonolus/core';
import { toServerItemDetails } from '../../models/server/items/details.js';
import { handleError } from '../error.js';
export const createDefaultServerItemDetailsHandler = (group) => ({ itemName }) => {
    const item = group.items.find(({ name }) => name === itemName);
    if (!item)
        return 404;
    const index = group.items.indexOf(item);
    return {
        item,
        description: item.description,
        actions: group.actions,
        hasCommunity: false,
        leaderboards: [],
        sections: [
            {
                title: { en: Text.Recommended },
                icon: Icon.Star,
                itemType: group.type,
                items: group.items.slice(index + 1, index + 6),
            },
        ],
    };
};
export const createServerItemDetailsRouteHandler = (sonolus, group, toItem) => async ({ req, res, localize, ctx }) => {
    const itemName = req.params.itemName;
    if (!itemName) {
        res.status(400).end();
        return;
    }
    const response = await group.detailsHandler({ ...ctx, itemName });
    if (handleError(response, res, localize))
        return;
    res.json(toServerItemDetails(sonolus, localize, toItem, response, group.searches, group.actions));
};
