// @sonolus/express/dist/routes/items/list.js
import { parseServerSearchesValue } from '../../models/server/forms/value.js';
import { toServerItemList } from '../../models/server/items/list.js';
import { extractString } from '../../utils/extract.js';
import { handleError } from '../error.js';
export const createDefaultServerItemListHandler = (group, filter) => ({ search, page }) => {
    const items = filter(group.items, search.type === 'quick' ? `${search.options.keywords}` : '');
    return {
        searches: group.searches,
        ...paginateItems(items, page),
    };
};
export const createServerItemListRouteHandler = (sonolus, group, toItem) => async ({ req, res, localize, ctx }) => {
    const response = await group.listHandler({
        ...ctx,
        search: parseServerSearchesValue(req.query, group.searches),
        page: +(req.query.page ?? '') || 0,
        cursor: extractString(req.query.cursor),
    });
    if (handleError(response, res, localize))
        return;
    res.json(toServerItemList(sonolus, localize, toItem, response, group.searches));
};
export const paginateItems = (items, page, perPage = 20) => ({
    pageCount: Math.ceil(items.length / perPage),
    items: items.slice(page * perPage, (page + 1) * perPage),
});
