// @sonolus/express/dist/sonolus/itemGroup.js
import { createServerItemCommunityCommentListRouteHandler, } from '../routes/items/community/comments/list.js';
import { createServerPreUploadItemCommunityCommentActionRouteHandler, } from '../routes/items/community/comments/preUpload.js';
import { createServerSubmitItemCommunityCommentActionRouteHandler, } from '../routes/items/community/comments/submit.js';
import { createServerUploadItemCommunityCommentActionRouteHandler, } from '../routes/items/community/comments/upload.js';
import { createServerItemCommunityInfoRouteHandler, } from '../routes/items/community/info.js';
import { createServerPreUploadItemCommunityActionRouteHandler, } from '../routes/items/community/preUpload.js';
import { createServerSubmitItemCommunityActionRouteHandler, } from '../routes/items/community/submit.js';
import { createServerUploadItemCommunityActionRouteHandler, } from '../routes/items/community/upload.js';
import { createServerCreateItemRouteHandler, } from '../routes/items/create.js';
import { createDefaultServerItemDetailsHandler, createServerItemDetailsRouteHandler, } from '../routes/items/details.js';
import { createDefaultServerItemInfoHandler, createServerItemInfoRouteHandler, } from '../routes/items/info.js';
import { createServerItemLeaderboardDetailsRouteHandler, } from '../routes/items/leaderboards/details.js';
import { createServerItemLeaderboardRecordDetailsRouteHandler, } from '../routes/items/leaderboards/records/details.js';
import { createServerItemLeaderboardRecordListRouteHandler, } from '../routes/items/leaderboards/records/list.js';
import { createDefaultServerItemListHandler, createServerItemListRouteHandler, } from '../routes/items/list.js';
import { createServerPreUploadItemRouteHandler, } from '../routes/items/preUpload.js';
import { createServerPreUploadItemActionRouteHandler, } from '../routes/items/preUploadAction.js';
import { createServerSubmitItemActionRouteHandler, } from '../routes/items/submitAction.js';
import { createServerUploadItemRouteHandler, } from '../routes/items/upload.js';
import { createServerUploadItemActionRouteHandler, } from '../routes/items/uploadAction.js';
export class SonolusItemGroup {
    constructor(sonolus, type, options = {}, toItem, filter) {
        this.type = type;
        this.items = [];
        this.creates = options.creates ?? {};
        this.searches = options.searches ?? {};
        this.actions = options.actions ?? {};
        this.infoHandler = createDefaultServerItemInfoHandler(sonolus, this);
        this.listHandler = createDefaultServerItemListHandler(this, filter);
        this.detailsHandler = createDefaultServerItemDetailsHandler(this);
        this.community = {
            actions: options.community?.actions ?? {},
            comment: {
                actions: options.community?.comment?.actions ?? {},
            },
        };
        this.leaderboard = {
            record: {},
        };
        this._infoRouteHandler = createServerItemInfoRouteHandler(sonolus, this);
        this._listRouteHandler = createServerItemListRouteHandler(sonolus, this, toItem);
        this._createRouteHandler = createServerCreateItemRouteHandler(this);
        this._preUploadRouteHandler = createServerPreUploadItemRouteHandler(this);
        this._uploadRouteHandler = createServerUploadItemRouteHandler(this);
        this._detailsRouteHandler = createServerItemDetailsRouteHandler(sonolus, this, toItem);
        this._submitActionRouteHandler = createServerSubmitItemActionRouteHandler(this);
        this._preUploadActionRouteHandler = createServerPreUploadItemActionRouteHandler(this);
        this._uploadActionRouteHandler = createServerUploadItemActionRouteHandler(this);
        this._communityInfoRouteHandler = createServerItemCommunityInfoRouteHandler(sonolus, this);
        this._communitySubmitRouteHandler = createServerSubmitItemCommunityActionRouteHandler(this);
        this._communityPreUploadRouteHandler =
            createServerPreUploadItemCommunityActionRouteHandler(this);
        this._communityUploadRouteHandler = createServerUploadItemCommunityActionRouteHandler(this);
        this._communityCommentListRouteHandler = createServerItemCommunityCommentListRouteHandler(sonolus, this);
        this._communityCommentSubmitRouteHandler =
            createServerSubmitItemCommunityCommentActionRouteHandler(this);
        this._communityCommentPreUploadRouteHandler =
            createServerPreUploadItemCommunityCommentActionRouteHandler(this);
        this._communityCommentUploadRouteHandler =
            createServerUploadItemCommunityCommentActionRouteHandler(this);
        this._leaderboardDetailsRouteHandler = createServerItemLeaderboardDetailsRouteHandler(sonolus, this);
        this._leaderboardRecordListRouteHandler = createServerItemLeaderboardRecordListRouteHandler(sonolus, this);
        this._leaderboardRecordDetailsRouteHandler =
            createServerItemLeaderboardRecordDetailsRouteHandler(sonolus, this);
    }
}
