// @sonolus/express/dist/sonolus/sonolus.js
import { localize, hash as sonolusHash, version } from '@sonolus/core';
import express from 'express';
import multer from 'multer';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { toBackgroundItem } from '../models/items/background.js';
import { toEffectItem } from '../models/items/effect.js';
import { toEngineItem } from '../models/items/engine.js';
import { toLevelItem } from '../models/items/level.js';
import { toParticleItem } from '../models/items/particle.js';
import { toPlaylistItem } from '../models/items/playlist.js';
import { toPostItem } from '../models/items/post.js';
import { toReplayItem } from '../models/items/replay.js';
import { toRoomItem } from '../models/items/room.js';
import { toSkinItem } from '../models/items/skin.js';
import { toUserItem } from '../models/items/user.js';
import { parseRawServerOptionsValue, parseServerOptionsValue, } from '../models/server/options/value.js';
import { createServerAuthenticateRouteHandler, } from '../routes/authenticate.js';
import { handleError } from '../routes/error.js';
import { filterUsers } from '../routes/index.js';
import { createDefaultServerInfoHandler, createServerInfoRouteHandler, } from '../routes/info.js';
import { filterBackgrounds } from '../routes/items/filters/background.js';
import { filterEffects } from '../routes/items/filters/effect.js';
import { filterEngines } from '../routes/items/filters/engine.js';
import { filterLevels } from '../routes/items/filters/level.js';
import { filterParticles } from '../routes/items/filters/particle.js';
import { filterPlaylists } from '../routes/items/filters/playlist.js';
import { filterPosts } from '../routes/items/filters/post.js';
import { filterReplays } from '../routes/items/filters/replay.js';
import { filterRooms } from '../routes/items/filters/room.js';
import { filterSkins } from '../routes/items/filters/skin.js';
import { databaseSchema } from '../schemas/database/database.js';
import { extractString } from '../utils/extract.js';
import { parse } from '../utils/json.js';
import { SonolusItemGroup } from './itemGroup.js';
import { SonolusLevelResult } from './levelResult.js';
import { SonolusMultiplayer } from './multiplayer.js';
const itemTypes = [
    ['post', 'posts', toPostItem, filterPosts],
    ['playlist', 'playlists', toPlaylistItem, filterPlaylists],
    ['level', 'levels', toLevelItem, filterLevels],
    ['skin', 'skins', toSkinItem, filterSkins],
    ['background', 'backgrounds', toBackgroundItem, filterBackgrounds],
    ['effect', 'effects', toEffectItem, filterEffects],
    ['particle', 'particles', toParticleItem, filterParticles],
    ['engine', 'engines', toEngineItem, filterEngines],
    ['replay', 'replays', toReplayItem, filterReplays],
    ['room', 'rooms', toRoomItem, filterRooms],
    ['user', 'users', toUserItem, filterUsers],
];
export class Sonolus {
    constructor(options = {}) {
        this.address = options.address;
        this.fallbackLocale = options.fallbackLocale ?? 'en';
        this.configuration = {
            options: options.configuration?.options ?? {},
        };
        this.router = express.Router();
        this.title = {};
        this.serverInfoHandler = createDefaultServerInfoHandler(this);
        this.multiplayer = new SonolusMultiplayer(this, () => this.room.creates);
        this.levelResult = new SonolusLevelResult(options.levelResult);
        for (const [type, , toItem, filter] of itemTypes) {
            this[type] = new SonolusItemGroup(this, type, options[type], toItem, filter);
        }
        this._installRoutes(options.upload ?? {});
    }
    load(path) {
        const dbPath = resolve(path, 'db.json');
        const repositoryPath = resolve(path, 'repository');
        const db = parse(readFileSync(dbPath, 'utf8'), databaseSchema);
        if (!db)
            throw new Error(`Invalid database: path=${dbPath}`);
        this.title = db.info.title;
        if (db.info.description)
            this.description = db.info.description;
        if (db.info.banner)
            this.banner = db.info.banner;
        this.post.items.push(...db.posts);
        this.playlist.items.push(...db.playlists);
        this.level.items.push(...db.levels);
        this.skin.items.push(...db.skins);
        this.background.items.push(...db.backgrounds);
        this.effect.items.push(...db.effects);
        this.particle.items.push(...db.particles);
        this.engine.items.push(...db.engines);
        this.replay.items.push(...db.replays);
        this.router.use('/sonolus/repository', express.static(repositoryPath));
    }
    add(data, hash) {
        hash ??= sonolusHash(typeof data === 'string' ? readFileSync(data) : data);
        const url = `/sonolus/repository/${hash}`;
        if (typeof data === 'string') {
            const path = resolve(data);
            this.router.get(url, (req, res) => {
                res.sendFile(path);
            });
        }
        else {
            this.router.get(url, (req, res) => {
                res.send(data);
            });
        }
        return { hash, url };
    }
    localize(text, locale) {
        return localize(text, locale, this.fallbackLocale);
    }
    _installRoutes(upload) {
        const uploader = multer(upload.options).array('files', upload.maxCount);
        this._post('/authenticate', createServerAuthenticateRouteHandler(this));
        this._get('/info', createServerInfoRouteHandler(this));
        this._post('/rooms/create', this.multiplayer['_createRouteHandler']);
        this._post('/rooms/:itemName', this.multiplayer['_joinRouteHandler']);
        this._get('/levels/result/info', this.levelResult['_infoRouteHandler']);
        this._post('/levels/result/submit', this.levelResult['_submitRouteHandler']);
        this._upload('/levels/result/upload', this.levelResult['_preUploadRouteHandler'], uploader, this.levelResult['_uploadRouteHandler']);
        for (const [type, path] of itemTypes) {
            this._get(`/${path}/info`, this[type]['_infoRouteHandler']);
            this._get(`/${path}/list`, this[type]['_listRouteHandler']);
            this._post(`/${path}/create`, this[type]['_createRouteHandler']);
            this._upload(`/${path}/upload`, this[type]['_preUploadRouteHandler'], uploader, this[type]['_uploadRouteHandler']);
            this._get(`/${path}/:itemName`, this[type]['_detailsRouteHandler']);
            this._post(`/${path}/:itemName/submit`, this[type]['_submitActionRouteHandler']);
            this._upload(`/${path}/:itemName/upload`, this[type]['_preUploadActionRouteHandler'], uploader, this[type]['_uploadActionRouteHandler']);
            this._get(`/${path}/:itemName/community/info`, this[type]['_communityInfoRouteHandler']);
            this._post(`/${path}/:itemName/community/submit`, this[type]['_communitySubmitRouteHandler']);
            this._upload(`/${path}/:itemName/community/upload`, this[type]['_communityPreUploadRouteHandler'], uploader, this[type]['_communityUploadRouteHandler']);
            this._get(`/${path}/:itemName/community/comments/list`, this[type]['_communityCommentListRouteHandler']);
            this._post(`/${path}/:itemName/community/comments/:commentName/submit`, this[type]['_communityCommentSubmitRouteHandler']);
            this._upload(`/${path}/:itemName/community/comments/:commentName/upload`, this[type]['_communityCommentPreUploadRouteHandler'], uploader, this[type]['_communityCommentUploadRouteHandler']);
            this._get(`/${path}/:itemName/leaderboards/:leaderboardName`, this[type]['_leaderboardDetailsRouteHandler']);
            this._get(`/${path}/:itemName/leaderboards/:leaderboardName/records/list`, this[type]['_leaderboardRecordListRouteHandler']);
            this._get(`/${path}/:itemName/leaderboards/:leaderboardName/records/:recordName`, this[type]['_leaderboardRecordDetailsRouteHandler']);
        }
    }
    _get(path, handler) {
        this.router.get(`/sonolus${path}`, this._toMiddleware(handler));
    }
    _post(path, handler) {
        this.router.post(`/sonolus${path}`, express.raw({ type: 'application/json' }), this._toMiddleware(handler));
    }
    _upload(path, preHandler, uploader, handler) {
        this.router.post(`/sonolus${path}`, this._toMiddleware(preHandler), uploader, this._toMiddleware(handler));
    }
    _toMiddleware(handler) {
        return async (req, res, next) => {
            try {
                res.set('Sonolus-Version', version.sonolus);
                const localization = extractString(req.query.localization) ?? '';
                const localize = (text) => this.localize(text, localization);
                const options = parseServerOptionsValue(req.query, this.configuration.options);
                const rawOptions = parseRawServerOptionsValue(req.query, this.configuration.options);
                const session = extractString(req.headers['sonolus-session']);
                const ctx = { session, localization, options, rawOptions };
                if (this.sessionHandler) {
                    const result = await this.sessionHandler(ctx);
                    if (handleError(result, res, localize))
                        return;
                }
                await handler({ req, res, next, localize, ctx });
            }
            catch (error) {
                console.error(error);
                res.status(500).end();
            }
        };
    }
}
