// @sonolus/express/dist/models/items/engine.js
import { toTags } from '../tag.js';
import { toBackgroundItem } from './background.js';
import { toEffectItem } from './effect.js';
import { getItem } from './item.js';
import { toParticleItem } from './particle.js';
import { toSkinItem } from './skin.js';
import { toUserItem } from './user.js';
export const toEngineItem = (sonolus, localize, item) => ({
    name: item.name,
    source: sonolus.address,
    version: item.version,
    title: localize(item.title),
    subtitle: localize(item.subtitle),
    author: localize(item.author),
    authorUser: item.authorUser
        ? toUserItem(sonolus, localize, getItem(sonolus.user.items, item.authorUser, `Engine/${item.name}`, '.authorUser'))
        : undefined,
    tags: toTags(localize, item.tags),
    skin: toSkinItem(sonolus, localize, getItem(sonolus.skin.items, item.skin, `Engine/${item.name}`, '.skin')),
    background: toBackgroundItem(sonolus, localize, getItem(sonolus.background.items, item.background, `Engine/${item.name}`, '.background')),
    effect: toEffectItem(sonolus, localize, getItem(sonolus.effect.items, item.effect, `Engine/${item.name}`, '.effect')),
    particle: toParticleItem(sonolus, localize, getItem(sonolus.particle.items, item.particle, `Engine/${item.name}`, '.particle')),
    thumbnail: item.thumbnail,
    playData: item.playData,
    watchData: item.watchData,
    previewData: item.previewData,
    tutorialData: item.tutorialData,
    rom: item.rom,
    configuration: item.configuration,
});
