// @sonolus/express/dist/models/items/level.js
import { toTags } from '../tag.js';
import { toBackgroundItem } from './background.js';
import { toEffectItem } from './effect.js';
import { toEngineItem } from './engine.js';
import { getItem } from './item.js';
import { toParticleItem } from './particle.js';
import { toSkinItem } from './skin.js';
import { toUserItem } from './user.js';
export const toLevelItem = (sonolus, localize, item) => ({
    name: item.name,
    source: sonolus.address,
    version: item.version,
    rating: item.rating,
    engine: toEngineItem(sonolus, localize, getItem(sonolus.engine.items, item.engine, `Level/${item.name}`, '.engine')),
    useSkin: toUseItem(sonolus, localize, toSkinItem, item.useSkin, sonolus.skin.items, `Level/${item.name}`, '.useSkin.item'),
    useBackground: toUseItem(sonolus, localize, toBackgroundItem, item.useBackground, sonolus.background.items, `Level/${item.name}`, '.useBackground.item'),
    useEffect: toUseItem(sonolus, localize, toEffectItem, item.useEffect, sonolus.effect.items, `Level/${item.name}`, '.useEffect.item'),
    useParticle: toUseItem(sonolus, localize, toParticleItem, item.useParticle, sonolus.particle.items, `Level/${item.name}`, '.useParticle.item'),
    title: localize(item.title),
    artists: localize(item.artists),
    author: localize(item.author),
    authorUser: item.authorUser
        ? toUserItem(sonolus, localize, getItem(sonolus.user.items, item.authorUser, `Level/${item.name}`, '.authorUser'))
        : undefined,
    tags: toTags(localize, item.tags),
    cover: item.cover,
    bgm: item.bgm,
    preview: item.preview,
    data: item.data,
});
export const toUseItem = (sonolus, localize, toItem, useItem, items, parent, path) => useItem.useDefault
    ? {
        useDefault: true,
    }
    : {
        useDefault: false,
        item: toItem(sonolus, localize, getItem(items, useItem.item, parent, path)),
    };
