// @sonolus/express/dist/models/server/forms/value.js
import { parseRawServerOptionsValue, parseServerOptionsValue, serializeRawServerOptionValue, } from '../options/value.js';
export const parseRawServerFormValue = (value, type, form) => ({
    type,
    rawOptions: parseRawServerOptionsValue(value, form.options),
});
export const parseServerFormValue = (value, type, form) => ({
    type,
    options: parseServerOptionsValue(value, form.options),
    rawOptions: parseRawServerOptionsValue(value, form.options),
});
export const serializeRawServerFormValue = (value, type, form) => new URLSearchParams([
    ['type', type],
    ...Object.entries(form.options)
        .map(([key, option]) => {
        const rawValue = value.rawOptions[key];
        if (rawValue === undefined)
            return;
        return [key, serializeRawServerOptionValue(rawValue, option)];
    })
        .filter((kvp) => kvp !== undefined),
]).toString();
export const parseServerFormsValue = (value, forms) => {
    const type = `${value.type}`;
    const form = forms[type];
    if (!form)
        return;
    return parseServerFormValue(value, type, form);
};
export const serializeRawServerFormsValue = (value, forms) => 
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
serializeRawServerFormValue(value, value.type, forms[value.type]);
const quickForm = {
    title: {},
    requireConfirmation: false,
    options: {
        keywords: {
            name: {},
            required: false,
            type: 'text',
            placeholder: {},
            def: '',
            limit: 0,
            shortcuts: [],
        },
    },
};
export const parseServerSearchesValue = (value, searches) => {
    const type = `${value.type}`;
    if (type === 'quick')
        return parseServerFormValue(value, 'quick', quickForm);
    return parseServerFormsValue(value, searches) ?? parseServerFormValue({}, 'quick', quickForm);
};
