// @sonolus/express/dist/routes/info.js
import { toServerInfo } from '../models/server/info.js';
import { handleError } from './error.js';
export const createDefaultServerInfoHandler = (sonolus) => () => ({
    title: sonolus.title,
    description: sonolus.description,
    buttons: [
        ...[
            'post',
            'playlist',
            'level',
            'replay',
            'skin',
            'background',
            'effect',
            'particle',
            'engine',
        ]
            .filter((type) => sonolus[type].items.length)
            .map((type) => ({ type })),
        { type: 'configuration' },
    ],
    configuration: {
        options: sonolus.configuration.options,
    },
    banner: sonolus.banner,
});
export const createServerInfoRouteHandler = (sonolus) => async ({ res, localize, ctx }) => {
    const response = await sonolus.serverInfoHandler(ctx);
    if (handleError(response, res, localize))
        return;
    res.json(toServerInfo(localize, response, sonolus.configuration.options));
};
