// @sonolus/express/dist/models/server/items/section.js
import { toBackgroundItem } from '../../items/background.js';
import { toEffectItem } from '../../items/effect.js';
import { toEngineItem } from '../../items/engine.js';
import { toItems } from '../../items/item.js';
import { toLevelItem } from '../../items/level.js';
import { toParticleItem } from '../../items/particle.js';
import { toPlaylistItem } from '../../items/playlist.js';
import { toPostItem } from '../../items/post.js';
import { toReplayItem } from '../../items/replay.js';
import { toRoomItem } from '../../items/room.js';
import { toSkinItem } from '../../items/skin.js';
import { toUserItem } from '../../items/user.js';
import { toServerForm } from '../forms/form.js';
import { serializeRawServerFormsValue } from '../forms/value.js';
const toItemByType = {
    post: toPostItem,
    playlist: toPlaylistItem,
    level: toLevelItem,
    skin: toSkinItem,
    background: toBackgroundItem,
    effect: toEffectItem,
    particle: toParticleItem,
    engine: toEngineItem,
    replay: toReplayItem,
    room: toRoomItem,
    user: toUserItem,
};
export const toServerItemSection = (sonolus, localize, section, searches) => ({
    title: localize(section.title),
    icon: section.icon,
    description: section.description && localize(section.description),
    help: section.help && localize(section.help),
    itemType: section.itemType,
    items: toItems(sonolus, localize, toItemByType[section.itemType], section.items),
    search: section.search?.form
        ? toServerForm(localize, section.search.value.type, section.search.form === true
            ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                searches[section.search.value.type]
            : section.search.form)
        : undefined,
    searchValues: section.search && serializeRawServerFormsValue(section.search.value, searches),
});
export const toServerItemSections = (sonolus, localize, sections, searches) => sections.map((section) => toServerItemSection(sonolus, localize, section, searches));
