// @sonolus/express/dist/models/server/options/collectionItem.js
import { backgroundItemSchema } from '../../../schemas/server/items/background.js';
import { effectItemSchema } from '../../../schemas/server/items/effect.js';
import { engineItemSchema } from '../../../schemas/server/items/engine.js';
import { levelItemSchema } from '../../../schemas/server/items/level.js';
import { particleItemSchema } from '../../../schemas/server/items/particle.js';
import { playlistItemSchema } from '../../../schemas/server/items/playlist.js';
import { postItemSchema } from '../../../schemas/server/items/post.js';
import { replayItemSchema } from '../../../schemas/server/items/replay.js';
import { roomItemSchema } from '../../../schemas/server/items/room.js';
import { skinItemSchema } from '../../../schemas/server/items/skin.js';
import { userItemSchema } from '../../../schemas/server/items/user.js';
import { parse } from '../../../utils/json.js';
const schemas = {
    post: postItemSchema,
    playlist: playlistItemSchema,
    level: levelItemSchema,
    skin: skinItemSchema,
    background: backgroundItemSchema,
    effect: effectItemSchema,
    particle: particleItemSchema,
    engine: engineItemSchema,
    replay: replayItemSchema,
    room: roomItemSchema,
    user: userItemSchema,
};
export const parseRawServerCollectionItemOptionValue = (value, option) => {
    if (typeof value !== 'string')
        return;
    return parse(value, schemas[option.itemType]);
};
export const normalizeServerCollectionItemOptionValue = (value) => value;
export const serializeServerCollectionItemOptionValue = (value) => JSON.stringify(value);
export const toServerCollectionItemOption = (localize, query, option) => ({
    query,
    name: localize(option.name),
    description: option.description && localize(option.description),
    required: option.required,
    type: option.type,
    itemType: option.itemType,
});
