// @sonolus/express/dist/models/server/options/option.js
import { toServerCollectionItemOption } from './collectionItem.js';
import { toServerFileOption } from './file.js';
import { toServerMultiOption } from './multi.js';
import { toServerSelectOption } from './select.js';
import { toServerServerItemOption } from './serverItem.js';
import { toServerServerItemsOption } from './serverItems.js';
import { toServerSliderOption } from './slider.js';
import { toServerTextOption } from './text.js';
import { toServerTextAreaOption } from './textArea.js';
import { toServerToggleOption } from './toggle.js';
export const toServerOption = (localize, query, option) => ({
    text: toServerTextOption,
    textArea: toServerTextAreaOption,
    slider: toServerSliderOption,
    toggle: toServerToggleOption,
    select: toServerSelectOption,
    multi: toServerMultiOption,
    serverItem: toServerServerItemOption,
    serverItems: toServerServerItemsOption,
    collectionItem: toServerCollectionItemOption,
    file: toServerFileOption,
})[option.type](localize, query, option);
export const toServerOptions = (localize, pick, options) => Object.entries(pick)
    .map(([type, value]) => 
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
value && toServerOption(localize, type, value === true ? options[type] : value))
    .filter((form) => !!form);
