// @sonolus/express/dist/models/server/options/value.js
import { normalizeServerCollectionItemOptionValue, parseRawServerCollectionItemOptionValue, serializeServerCollectionItemOptionValue, } from './collectionItem.js';
import { normalizeServerFileOptionValue, parseRawServerFileOptionValue, serializeServerFileOptionValue, } from './file.js';
import { normalizeServerMultiOptionValue, parseRawServerMultiOptionValue, serializeServerMultiOptionValue, } from './multi.js';
import { normalizeServerSelectOptionValue, parseRawServerSelectOptionValue, serializeServerSelectOptionValue, } from './select.js';
import { normalizeServerServerItemOptionValue, parseRawServerServerItemOptionValue, serializeServerServerItemOptionValue, } from './serverItem.js';
import { normalizeServerServerItemsOptionValue, parseRawServerServerItemsOptionValue, serializeServerServerItemsOptionValue, } from './serverItems.js';
import { normalizeServerSliderOptionValue, parseRawServerSliderOptionValue, serializeServerSliderOptionValue, } from './slider.js';
import { normalizeServerTextOptionValue, parseRawServerTextOptionValue, serializeServerTextOptionValue, } from './text.js';
import { normalizeServerTextAreaOptionValue, parseRawServerTextAreaOptionValue, serializeServerTextAreaOptionValue, } from './textArea.js';
import { normalizeServerToggleOptionValue, parseRawServerToggleOptionValue, serializeServerToggleOptionValue, } from './toggle.js';
const parseRawServerOptionValueByType = {
    text: parseRawServerTextOptionValue,
    textArea: parseRawServerTextAreaOptionValue,
    slider: parseRawServerSliderOptionValue,
    toggle: parseRawServerToggleOptionValue,
    select: parseRawServerSelectOptionValue,
    multi: parseRawServerMultiOptionValue,
    serverItem: parseRawServerServerItemOptionValue,
    serverItems: parseRawServerServerItemsOptionValue,
    collectionItem: parseRawServerCollectionItemOptionValue,
    file: parseRawServerFileOptionValue,
};
export const parseRawServerOptionValue = (value, option) => parseRawServerOptionValueByType[option.type](value, option);
const normalizeServerOptionValueByType = {
    text: normalizeServerTextOptionValue,
    textArea: normalizeServerTextAreaOptionValue,
    slider: normalizeServerSliderOptionValue,
    toggle: normalizeServerToggleOptionValue,
    select: normalizeServerSelectOptionValue,
    multi: normalizeServerMultiOptionValue,
    serverItem: normalizeServerServerItemOptionValue,
    serverItems: normalizeServerServerItemsOptionValue,
    collectionItem: normalizeServerCollectionItemOptionValue,
    file: normalizeServerFileOptionValue,
};
export const parseServerOptionValue = (value, option) => normalizeServerOptionValueByType[option.type](parseRawServerOptionValue(value, option), option);
export const serializeRawServerOptionValue = (value, option) => ({
    text: serializeServerTextOptionValue,
    textArea: serializeServerTextAreaOptionValue,
    slider: serializeServerSliderOptionValue,
    toggle: serializeServerToggleOptionValue,
    select: serializeServerSelectOptionValue,
    multi: serializeServerMultiOptionValue,
    serverItem: serializeServerServerItemOptionValue,
    serverItems: serializeServerServerItemsOptionValue,
    collectionItem: serializeServerCollectionItemOptionValue,
    file: serializeServerFileOptionValue,
})[option.type](value);
export const parseRawServerOptionsValue = (value, options) => Object.fromEntries(Object.entries(options).map(([key, option]) => [
    key,
    parseRawServerOptionValue(value[key], option),
]));
export const parseServerOptionsValue = (value, options) => Object.fromEntries(Object.entries(options).map(([key, option]) => [
    key,
    parseServerOptionValue(value[key], option),
]));
